### PR TITLE
Neptune logger

### PR DESCRIPTION
**Pull request by AleksanderWWW**
# What does this PR do?

Adds Neptune logger support

# What issue(s) does this change relate to?

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by AleksanderWWW** [comment]
@eracah is there any example script using logger's artifact logging functionality?

File: composer/loggers/neptune_logger.py
Comment (SiddhantSadangi) [inline composer/loggers/neptune_logger.py:148]: This should depend on whether `upload_artifacts` is set to `True`

```suggestion
        return (self._enabled and self._upload_artifacts)
```

File: composer/loggers/neptune_logger.py
Comment (SiddhantSadangi) [inline composer/loggers/neptune_logger.py:230]: Do we explicitly need to call `sync()` before `stop()`?

File: composer/loggers/neptune_logger.py
Comment (SiddhantSadangi) [inline composer/loggers/neptune_logger.py:39]: The name `log_artifacts` might be misleading for Neptune users as tracking artifacts is different from uploading them to Neptune.

```suggestion
        upload_artifacts (bool, optional): Whether the logger should upload artifacts to Neptune.ai.
```

File: composer/loggers/neptune_logger.py
Comment (SiddhantSadangi) [inline composer/loggers/neptune_logger.py:41]:
```suggestion
        base_namespace (str, optional): The name of the base namespace to log the metadata to. (default: "training")
```

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

File: composer/loggers/neptune_logger.py
Comment (AleksanderWWW) [reply composer/loggers/neptune_logger.py:230]: Afaik `stop()` doesn't sync the data - on the contrary, I think it actually kills a synchronization thread

**Comment by AleksanderWWW** [comment]
Thank you! I see that smoke tests are failing because the `neptune` package doesn't get installed. How shall I fix it?

Screenshot from 2023-09-22 17-10-10

**Comment by mvpatel2000** [comment]
> Thank you! I see that smoke tests are failing because the `neptune` package doesn't get installed. How shall I fix it? 

@AleksanderWWW can you please in-line all `import neptune` into each function? For example, see the wandb logger here: https://github.com/mosaicml/composer/blob/dev/composer/loggers/wandb_logger.py#L66-L72. Otherwise, `import Composer` fails unless a user has neptune installed which is bad as it is an optional dependency

**Comment by AleksanderWWW** [comment]
> > Thank you! I see that smoke tests are failing because the `neptune` package doesn't get installed. How shall I fix it?
> 
> @AleksanderWWW can you please in-line all `import neptune` into each function? For example, see the wandb logger here: https://github.com/mosaicml/composer/blob/dev/composer/loggers/wandb_logger.py#L66-L72. Otherwise, `import Composer` fails unless a user has neptune installed which is bad as it is an optional dependency

Hey, done :) Thank you!

**Comment by mvpatel2000** [comment]
@AleksanderWWW  is this PR ready for review / can we mark as not a draft?

File: tests/loggers/test_neptune_logger.py
Comment (dakinggg) [inline tests/loggers/test_neptune_logger.py:81]: if you are expecting this test to run on multiple gpus, you'll need to make it with `@pytest.mark.gpu` and `@pytest.mark.world_size(2)` (or `@pytest.mark.world_size(1, 2)` if you want it to do both 1 gpu and 2 gpus)

File: examples/imagenet/train_resnet_imagenet1k_neptune.py
Comment (dakinggg) [inline examples/imagenet/train_resnet_imagenet1k_neptune.py:1]: this example is a bit out of place I think, and it would be better to just augment a few places in the docs that we describe how to use loggers to add neptune. so https://github.com/mosaicml/composer/blob/d3d3a4e99ce1288b7ee1ef1e1b8a59cf61c31fdc/docs/source/trainer/logging.rst#L43 and https://github.com/mosaicml/composer/blob/d3d3a4e99ce1288b7ee1ef1e1b8a59cf61c31fdc/docs/source/trainer/file_uploading.rst#L91 at a glance. Or, rather than adding a new script just for neptune, augment the existing example to accept neptune logger args. @mvpatel2000 wdyt?

Comment (AleksanderWWW) [reply examples/imagenet/train_resnet_imagenet1k_neptune.py:1]: Ah apologies, I added it chiefly to give my colleagues a very minimal example to understand how this integration is supposed to work. I will delete it after their review

**Comment by mvpatel2000** [comment]
> @mvpatel2000 We could, but even if it is approved from your side, I'd like to also have an approval from my team if it's not a problem :)

Let us know when you are ready in that case :)

File: composer/loggers/neptune_logger.py
Comment (SiddhantSadangi) [inline composer/loggers/neptune_logger.py:112]: We don't need to use tags here

```suggestion
        if not self._rank_zero_only:
            self._base_handler["rank"] = dist.get_global_rank()

        if self._enabled:
            assert self._neptune_run is not None
            self._neptune_run['sys/name'] = state.run_name
```

**Review by normandy7 — COMMENTED** [review]
Proposing docstrings

File: composer/loggers/neptune_logger.py
Comment (Raalsky) [inline composer/loggers/neptune_logger.py:133]: Let's use `_ = logger` 😜 instead. `logger` is a reference so I'm not 100% sure what calling del here means.

File: composer/loggers/neptune_logger.py
Comment (Raalsky) [inline composer/loggers/neptune_logger.py:103]: A few other integrations showed us that it's quite unsafe to initialize a Run in class constructor. Let's made it lazy evaluated in `neptune_run` property. And let's just use a Run in a `debug` mode to prevent the whole logic of `if self._neptune_run is not None:...`

**Comment by SiddhantSadangi** [comment]
I am also getting the asynchronous operation processing error: `Error occurred during asynchronous operation processing: X-coordinates (step) must be strictly increasing for series attribute: training/metrics/metrics/train/MulticlassAccuracy`

Could you check why?

Code to reproduce: https://github.com/neptune-ai/examples/blob/eb1bf0205c526ca6ab4219fc405a18b2e4cd01a7/integrations-and-supported-tools/mosaicml-composer/notebooks/Neptune_MosaicML_Composer.ipynb

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]
I am also getting the asynchronous operation processing error: Error occurred during asynchronous operation processing: X-coordinates (step) must be strictly increasing for series attribute: training/metrics/metrics/train/MulticlassAccuracy

Could you check why?

Code to reproduce: https://github.com/neptune-ai/examples/blob/eb1bf0205c526ca6ab4219fc405a18b2e4cd01a7/integrations-and-supported-tools/mosaicml-composer/notebooks/Neptune_MosaicML_Composer.ipynb

File: composer/loggers/neptune_logger.py
Comment (normandy7) [inline composer/loggers/neptune_logger.py:138]: Docstring for `base_handler` property

```suggestion
    def base_handler(self):
        """Gets a handler for the base logging namespace.
    
        Use the handler to log extra metadata to the run and organize it under the base namespace (default: "training").
        You can operate on it like a run object: Access a path inside the handler and assign metadata to it with "=" or
        other [Neptune logging methods](https://docs.neptune.ai/logging/methods/).
    
        Example:
            >>> neptune_logger = NeptuneLogger()
            >>> trainer = Trainer(loggers=neptune_logger, ...)
            >>> trainer.fit()
            >>> neptune_logger.base_handler["some_metric"] = 1
            >>> trainer.close()
            Result: The value `1` is organized under "training/some_metric" inside the run.
        """
        return self.neptune_run[self._base_namespace]
```

File: composer/loggers/neptune_logger.py
Comment (Raalsky) [inline composer/loggers/neptune_logger.py:256]: We can probably just use `.extend` instead

File: composer/loggers/neptune_logger.py
Comment (Raalsky) [inline composer/loggers/neptune_logger.py:243]: We should probably transform images to match our RGB index.

**Review by Raalsky — APPROVED** [review]

**Review by normandy7 — APPROVED** [review]

**Comment by AleksanderWWW** [comment]
@mvpatel2000 Hey, it took a while, but we're ready for review :smile:

**Comment by mvpatel2000** [comment]
@eracah @dakinggg  can u take a look?

**Comment by mvpatel2000** [comment]
> @mvpatel2000 Hey, it took a while, but we're ready for review 😄

@AleksanderWWW looks like a bunch of tests fail, can you please address those?

**Comment by SiddhantSadangi** [comment]
@AleksanderWWW - this issue still exists: https://github.com/mosaicml/composer/pull/2447#pullrequestreview-1657091331

File: composer/loggers/neptune_logger.py
Comment (normandy7) [inline composer/loggers/neptune_logger.py:56]: We can remove the interactive formatting, as this is just mocked sample code that highlights Neptune-specific parts.

```suggestion
            from composer.loggers import NeptuneLogger
            neptune_logger = NeptuneLogger()
            trainer = Trainer(loggers=neptune_logger, ...)
            trainer.fit()

        More options:
            neptune_logger = NeptuneLogger(
                project="ml-team/classification",
                upload_artifacts=True,
                dependencies="infer",  # Neptune Run kwarg
            )
            trainer = Trainer(loggers=neptune_logger, ...)
            trainer.fit()
```

Comment (AleksanderWWW) [reply composer/loggers/neptune_logger.py]: Ah, didn't notice that before pushing the last commit. Let me correct that manually in a second

File: docs/source/trainer/logging.rst
Comment (Raalsky) [inline docs/source/trainer/logging.rst:50]: Can we update the `README.md` as well? There is a section about integrations.

File: docs/source/trainer/logging.rst
Comment (Raalsky) [inline docs/source/trainer/logging.rst:45]: Should we update `tests/callbacks/callback_settings.py` somehow?

File: composer/loggers/neptune_logger.py
Comment (Raalsky) [inline composer/loggers/neptune_logger.py:65]: Is it intentionally not to use lowercase here?

File: composer/loggers/neptune_logger.py
Comment (Raalsky) [inline composer/loggers/neptune_logger.py:78]: It's not going to work with `neptune-client<1.0.0`, right? Shouldn't we add support for it?

File: composer/loggers/neptune_logger.py
Comment (Raalsky) [inline composer/loggers/neptune_logger.py:177]: What about `neptune-client<1.0.0`?

File: docs/source/trainer/logging.rst
Comment (normandy7) [reply docs/source/trainer/logging.rst:50]: You can also leave it to me. Loggers are mentioned in `docs/source/trainer/logging.rst` too. Let me know, so we don't get our wires crossed.

Comment (AleksanderWWW) [reply docs/source/trainer/logging.rst:50]: @normandy7 I will then leave it in your more competent hands :rocket: :bow:

Comment (normandy7) [reply docs/source/trainer/logging.rst:50]: I failed to note that this PR already includes changes to `logging.rst` 🤦🏻‍♀️ We're just missing a mention of neptune.ai on line 27, but I'm unable to commit/suggest directly.

```
Biases <https://www.wandb.com/>`__, `MLflow <https://www.mlflow.org/docs/latest/index.html>`__, `CometML <https://www.comet.com/?utm_source=mosaicml&utm_medium=partner&utm_campaign=mosaicml_comet_integration>`__, and `neptune.ai <https://neptune.ai/>`__,
```

Comment (normandy7) [reply docs/source/trainer/logging.rst:50]: As for the README, updating line 108 to this should suffice:

```
- **Experiment tracking:** Weights and Biases, MLFlow, CometML, and neptune.ai — the choice is yours, easily log your data to your favorite platform.
```

I assumed there'd be more, but if the docstrings are now OK, as far as I can tell these neptune.ai mentions are all that's missing.

Comment (AleksanderWWW) [reply docs/source/trainer/logging.rst:50]: Got it. The last commit should cover that. Could you confirm?

Comment (normandy7) [reply docs/source/trainer/logging.rst:50]: Yep, looks good on my end.

**Review by normandy7 — APPROVED** [review]

File: composer/loggers/neptune_logger.py
Comment (Raalsky) [inline composer/loggers/neptune_logger.py:181]: Can we extract the logic below to a separate private method or a function?

**Review by Raalsky — APPROVED** [review]

File: composer/loggers/neptune_logger.py
Comment (AleksanderWWW) [reply composer/loggers/neptune_logger.py:181]: Done

**Review by Raalsky — APPROVED** [review]

**Comment by AleksanderWWW** [comment]
@mvpatel2000 @dakinggg @eracah Might I ask for your support/review?

**Comment by AleksanderWWW** [comment]
@SiddhantSadangi Seems that the bug you discovered should now be gone, as we made step logging more robust. Could you confirm?

**Review by SiddhantSadangi — APPROVED** [review]
LGTM 🚀 
Don't forget to delete `examples/imagenet/train_resnet_imagenet1k_neptune.py` :)

**Comment by normandy7** [comment]
@AleksanderWWW One last thing: we can update the text on line 49 in `composer.callbacks.ImageVisualizer.py` to

```
This callback only works with wandb and Neptune logging for now.
```

**Comment by mvpatel2000** [comment]
@AleksanderWWW  checking in if this is ready for review

**Comment by AleksanderWWW** [comment]
@mvpatel2000 Generally speaking it is, but we struggle with doctest issues that are hard to reproduce locally for some reason

**Comment by AleksanderWWW** [comment]
```
/usr/lib/python3/dist-packages/composer/loggers/neptune_logger.py:docstring of composer.loggers.neptune_logger.NeptuneLogger:34: ERROR: Unexpected indentation.
E       /usr/lib/python3/dist-packages/composer/loggers/neptune_logger.py:docstring of composer.loggers.neptune_logger.NeptuneLogger:37: WARNING: Block quote ends without a blank line; unexpected unindent.
```
@mvpatel2000 @normandy7 
Even after removing those "problematic" lines the doctest still fails in exactly the same places.
I compared our docs to wandb and honestly cannot see any divergence in the format.
No idea what to do with that.

**Comment by normandy7** [comment]
> ```
> /usr/lib/python3/dist-packages/composer/loggers/neptune_logger.py:docstring of composer.loggers.neptune_logger.NeptuneLogger:34: ERROR: Unexpected indentation.
> E       /usr/lib/python3/dist-packages/composer/loggers/neptune_logger.py:docstring of composer.loggers.neptune_logger.NeptuneLogger:37: WARNING: Block quote ends without a blank line; unexpected unindent.
> ```
> 
> @mvpatel2000 @normandy7 Even after removing those "problematic" lines the doctest still fails in exactly the same places. I compared our docs to wandb and honestly cannot see any divergence in the format. No idea what to do with that.

The main difference I see is that we don't use reST markup for the external links. I'll propose some markup changes in case that would help.

File: composer/loggers/neptune_logger.py
Comment (normandy7) [inline composer/loggers/neptune_logger.py:40]: @AleksanderWWW Let's try this

```suggestion
    """Log to `neptune.ai <https://neptune.ai/>`_.

    Args:
        project (str, optional): The name of your Neptune project,
            in the form "workspace-name/project-name". If you leave it empty, the
            ``NEPTUNE_PROJECT`` environment variable will be used.
        api_token (str, optional): Your Neptune API token.
            You can leave out this argument if you save your token to the
            ``NEPTUNE_API_TOKEN`` environment variable (recommended).
            You can find your API token in the user menu of the Neptune web app.
        rank_zero_only (bool, optional): Whether to log only on the rank-zero process.
            Default: ``True``.
        upload_artifacts (bool, optional): Whether the logger should upload artifacts to Neptune.
            Default: ``False``.
        base_namespace (str, optional): The name of the base namespace to log the metadata to.
            Default: "training".
        neptune_kwargs (Dict[str, Any], optional): Any additional keyword arguments to the
            ``neptune.init_run()`` function. For options, see the
            `Run API reference <https://docs.neptune.ai/api/neptune/#init_run>`_ in the
            Neptune docs.
```

File: composer/loggers/neptune_logger.py
Comment (normandy7) [inline composer/loggers/neptune_logger.py:66]: @AleksanderWWW -- are doctests still failing? (Not sure how to check this)

I'm noticing we have still Markdown in some places, which I can change to reST if the issues persist. If that doesn't help, we can try removing all of the `Examples` sections from the docstrings.

Comment (AleksanderWWW) [reply composer/loggers/neptune_logger.py:66]: Not sure, I've some troubles running the check locally and the workflow in GH actions requires approval from code owners :(

**Comment by mvpatel2000** [comment]
@AleksanderWWW can you resolve the merge conflicts and maybe I can take this PR over to sort out the unit tests + merge?

**Comment by AleksanderWWW** [comment]
@mvpatel2000 done

**Comment by mvpatel2000** [comment]
@AleksanderWWW it seems `os.environ["NEPTUNE_MODE"] = "offline"` is not enough and we still hit error: 
```
      File "/usr/lib/python3/dist-packages/composer/loggers/neptune_logger.py", line 151, in init
        self.base_handler['rank'] = dist.get_global_rank()
      File "/usr/lib/python3/dist-packages/composer/loggers/neptune_logger.py", line 146, in base_handler
        return self.neptune_run[self._base_namespace]
      File "/usr/lib/python3/dist-packages/composer/loggers/neptune_logger.py", line 120, in neptune_run
        self._neptune_run = Run(
      File "/usr/lib/python3/dist-packages/neptune/metadata_containers/run.py", line 408, in __init__
        super().__init__(
      File "/usr/lib/python3/dist-packages/neptune/metadata_containers/metadata_container.py", line 149, in __init__
        self._backend: NeptuneBackend = get_backend(mode=mode, api_token=api_token, proxies=proxies)
      File "/usr/lib/python3/dist-packages/neptune/internal/backends/factory.py", line 31, in get_backend
        return HostedNeptuneBackend(credentials=Credentials.from_token(api_token=api_token), proxies=proxies)
      File "/usr/lib/python3/dist-packages/neptune/internal/credentials.py", line 49, in from_token
        raise NeptuneMissingApiTokenException()
    neptune.exceptions.NeptuneMissingApiTokenException: 
```

can you provide guidance?

**Comment by mvpatel2000** [comment]
@AleksanderWWW  it seems all the unit tests, not just the doctests, fail because it complains there is no token even when on offline mode. Is this a bug on neptune's end?

**Comment by AleksanderWWW** [comment]
I've no idea why it should happen and cannot reproduce this scenario locally, but let's then try with setting the offline mode directly in the logger constructor

**Comment by AleksanderWWW** [comment]
> @AleksanderWWW - could you try setting "debug" mode instead of "offline"

@SiddhantSadangi 
Let's try that if setting the mode in constructor fails too

**Review by mvpatel2000 — APPROVED** [review]
@AleksanderWWW let's just remove the doctest for now. I can't seem to get it to cleanly pass, and I don't think it makes sense to hold up this PR any longer for it. I will go and fix it later on.

Thank you for your contribution!

**Comment by AleksanderWWW** [comment]
@mvpatel2000 Awesome! Thank you! Does that mean that it can be merged, or are we waiting for other reviewers?

File: composer/loggers/neptune_logger.py
Comment (dakinggg) [inline composer/loggers/neptune_logger.py:99]: This doesn't need a super init because the base class is ABC without an init, but if you want to leave it, prefer the syntax `super().__init__()`

**Review by dakinggg — APPROVED** [review]

File: composer/loggers/neptune_logger.py
Comment (AleksanderWWW) [reply composer/loggers/neptune_logger.py:99]: I just wanted PyCharm to stop complaining about lack of call to the super class init :laughing:

Comment (mvpatel2000) [reply composer/loggers/neptune_logger.py:99]: That's fine. I'll edit and we can merge.

**Pull request merged**
